File: ios/Sources/VaultSync/Presentation/ViewModels/DocumentDetailViewModel.swift
import SwiftUI
import Combine

@MainActor
public final class DocumentDetailViewModel: ObservableObject {
    @Published public var document: Document?
    @Published public var integrityResult: IntegrityResult?
    @Published public var isVerifying: Bool = false

    private let documentRepository: any DocumentRepository
    private let verifyIntegrityUseCase: VerifyIntegrityUseCase

    public init(
        documentRepository: any DocumentRepository,
        verifyIntegrityUseCase: VerifyIntegrityUseCase
    ) {
        self.documentRepository = documentRepository
        self.verifyIntegrityUseCase = verifyIntegrityUseCase
    }

    public func loadDocument(id: String) {
        Task {
            document = try? await documentRepository.getDocument(id: id)
            integrityResult = nil
        }
    }

    public func verifyIntegrity(id: String) {
        isVerifying = true
        Task {
            defer { isVerifying = false }
            integrityResult = try? await verifyIntegrityUseCase.execute(documentId: id)
        }
    }
}

File: android/app/src/main/java/com/vaultsync/ui/viewmodel/DocumentDetailViewModel.kt
package com.vaultsync.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vaultsync.domain.model.Document
import com.vaultsync.domain.model.IntegrityResult
import com.vaultsync.domain.repository.DocumentRepository
import com.vaultsync.domain.usecase.VerifyIntegrityUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class DocumentDetailViewModel(
    private val documentRepository: DocumentRepository,
    private val verifyIntegrityUseCase: VerifyIntegrityUseCase
) : ViewModel() {

    private val _document = MutableStateFlow<Document?>(null)
    val document: StateFlow<Document?> = _document.asStateFlow()

    private val _integrityResult = MutableStateFlow<IntegrityResult?>(null)
    val integrityResult: StateFlow<IntegrityResult?> = _integrityResult.asStateFlow()

    private val _isVerifying = MutableStateFlow(false)
    val isVerifying: StateFlow<Boolean> = _isVerifying.asStateFlow()

    fun loadDocument(id: String) {
        viewModelScope.launch {
            _document.value = documentRepository.getDocument(id)
            _integrityResult.value = null
        }
    }

    fun verifyIntegrity(id: String) {
        viewModelScope.launch {
            _isVerifying.value = true
            try {
                _integrityResult.value = verifyIntegrityUseCase(id)
            } finally {
                _isVerifying.value = false
            }
        }
    }
}
